### PR TITLE
feat: openapi migration

### PR DESCRIPTION
**Pull request by cpaulve-1A**
## Proposed change
Migrate the typescript sdk generator in the @ama-sdk/schematics to extends OpenApiGenerator instead of Swagger generator to cover swagger and open api specifications

## Related issues

- :bug: Fixes #(issue)
- :rocket: Feature #(issue)

### CODE REVIEW

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/TypescriptfetchGenerator.java [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-import io.swagger.models.properties.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.*;
 import java.io.File;
 
 public class TypescriptfetchGenerator extends AbstractTypeScriptClientCodegen {
 
+  private final Logger LOGGER = LoggerFactory.getLogger(TypescriptfetchGenerator.class);
```
Comment (cpaulve-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/TypescriptfetchGenerator.java:11]: not needed

File: packages/@ama-sdk/schematics/schematics/code-generator/swagger-java-generator/swagger-java.generator.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import * as path from 'node:path';
+import { spawn, SpawnOptions } from 'node:child_process';
+import { CodeGenerator } from '../code-generator';
+import { defaultOptions, JavaGeneratorTaskOptions } from './swagger-java.options';
+
+export class SwaggerJavaGenerator extends CodeGenerator<JavaGeneratorTaskOptions> {
+  protected generatorName = 'java-generator';
+
+  protected getDefaultOptions = () => defaultOptions;
+  protected runCodeGeneratorFactory(factoryOptions: { rootDirectory?: string } = {}) {
```
Comment (cpaulve-1A) [inline packages/@ama-sdk/schematics/schematics/code-generator/swagger-java-generator/swagger-java.generator.ts:10]: add tsdoc for all the protected/public methods

File: packages/@ama-sdk/generator-sdk/src/generators/core/templates/swagger-codegen-typescript/src/main/resources/META-INF/services/org.openapitools.codegen.CodegenConfig
Comment (cpaulve-1A) [inline packages/@ama-sdk/generator-sdk/src/generators/core/templates/swagger-codegen-typescript/src/main/resources/META-INF/services/org.openapitools.codegen.CodegenConfig:1]: todo: remove this file

File: packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      ...generatorOptions.specConfigPath ? ['-c', generatorOptions.specConfigPath] : [],
+      '-o', generatorOptions.outputPath
+    ];
+    return new Promise<void>((resolve, reject) => {
+      spawn(this.packageManager === 'npm' ? 'npx' : this.packageManager, args, spawnOptions)
+        .on('close', (code: number) => {
+          if (code === 0) {
+            resolve();
+          } else {
+            reject(new Error(`OpenApiGeneratorCli failed to run OpenApiGenerator with command 'run ${args.join(' ')}'`));
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts:57]:
```suggestion
            reject(new Error(`OpenApiGeneratorCli failed to run OpenApiGenerator with command '${args.join(' ')}'`));
```
as `run` is the 1st param of `args`

File: packages/@ama-sdk/schematics/schematics/ng-add/index.ts [hunk 4]
```diff
@@ -74,7 +74,7 @@ export const updatePackageJsonScripts: Rule = (tree, context) => {
             /@ama-sdk\/(schematics|generator-sdk):(core|shell|create|mock)/g,
             '@ama-sdk/schematics:typescript-$2'
           ) // Change typescript schematics name
-          .replaceAll('--swaggerSpecPath', '--swagger-spec-path') // Schematics arguments should be kebab-case
+          .replaceAll('--specPath', '--spec-path') // Schematics arguments should be kebab-case
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/ng-add/index.ts:77]:
```suggestion
          .replaceAll('--swaggerSpecPath', '--spec-path') // Schematics arguments should be kebab-case
```

File: packages/@ama-sdk/schematics/testing/MOCK_package.json [hunk 5]
```diff
@@ -1,11 +1,11 @@
 {
   "name": "fake-package",
   "scripts": {
-    "swagger:regen": "yo ./node_modules/@ama-sdk/generator-sdk/src/generators/java-client-core --swaggerSpecPath ./swagger-spec.yaml --swaggerConfigPath node_modules/@ama-sdk/generator-sdk/src/generators/java-client-core/templates/swagger-codegen-java-client/config/swagger-codegen-config.json --force",
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/testing/MOCK_package.json:4]: This file is use for test to apply the changes of ng-add on it 
so you should not modify it I think

File: packages/@ama-sdk/schematics/schematics/ng-add/index.ts (hunk 4 above)
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/ng-add/index.ts:77]: why when we support both swagger and open api spec?

File: packages/@ama-sdk/schematics/testing/MOCK_package.json (hunk 5 above)
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/testing/MOCK_package.json:4]: not if I want to rename the parameter

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/LambdaHelper.java [hunk 6]
```diff
@@ -516,7 +516,7 @@ public static String getFirstJsonMimeType(String[] mimeTypes) {
             return mime;
           }
         }
-        return mimeTypes[0];
+        return !mimeTypes[0].equals("") ? mimeTypes[0] : "application/json";
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/LambdaHelper.java:519]: Maybe we should let it empty and do not add the header if `""`

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/openapitools.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "$schema": "./node_modules/@openapitools/openapi-generator-cli/config.schema.json",
```
Comment (kpanot) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/openapitools.json:2]: please use github URL to allow yarn pnp support

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/package.json [hunk 8]
```diff
@@ -80,6 +80,7 @@
     "@ama-sdk/core": "~<%= sdkCoreVersion %>",
     "@o3r/eslint-config-otter": "^<%= sdkCoreVersion %>",
     "@o3r/eslint-plugin": "^<%= sdkCoreVersion %>",
+    "@openapitools/openapi-generator-cli": "~2.6.0",
```
Comment (kpanot) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/package.json:83]: Maybe more future proof to use the version from the peerDep of the repo

File: packages/@ama-sdk/schematics/schematics/code-generator/code-generator.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+   * Output path for the generated sdk
+   */
+  outputPath: string;
+};
+
+/**
+ * Expose angular schematics to run any code generator.
+ * Handle the task registration, configuration and the schematic's schedule.
+ *
+ * As is the CodeGenerator does not implement any actual code generation and needs to be extended to be functional
```
[the remark this answers is not in the document]
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/code-generator/code-generator.ts:28]:
```suggestion
 * As is, the CodeGenerator does not implement any actual code generation and needs to be extended to be functional
```

File: packages/@ama-sdk/schematics/schematics/code-generator/code-generator.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * Returns the schematic to register, configure and run your code generator task
+   *
+   * @param generatorOptions to configure the generator and the specification file to use
+   * @param factoryOptions to configure the generator execution context
+   * @param factoryOptions.rootDirectory directory where all your commands will be run - fallback to process.cwd
+   */
+  public getGeneratorRunSchematic(generatorOptions: Partial<T>, factoryOptions: { rootDirectory?: string } = {}) {
+    const scheduleTask = (tree: Tree, context: SchematicContext) => {
```
[the remark this answers is not in the document]
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/code-generator/code-generator.ts:102]: I need to schedule the task in a context, and this method is a rule factory. I am not sure how I am supposed to schedile the task without the rule context

File: packages/@ama-sdk/schematics/schematics/index.it.spec.ts [hunk 11]
```diff
@@ -54,6 +54,9 @@ function setupYarn(yarnVersion: string) {
   execSync('yarn config set unsafeHttpWhitelist localhost', execAppOptions);
   execSync('yarn config set nodeLinker pnp', execAppOptions);
   execSync('yarn config set enableMirror false', execAppOptions);
+  if (!existsSync(cacheFolderPath)) {
+    mkdirSync(cacheFolderPath);
```
[the remark this answers is not in the document]
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/index.it.spec.ts:58]: -p does not work well with windows, however I can use {recursive: true}

File: packages/@ama-sdk/schematics/schematics/typescript/core/index.ts [hunk 12]
```diff
@@ -70,10 +68,10 @@ export function ngGenerateTypescriptSDK(options: NgGenerateTypescriptSDKCoreSche
    * @param _context
    */
   const clearGeneratedCode = (tree: Tree, _context: SchematicContext) => {
```
[the remark this answers is not in the document]
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/typescript/core/index.ts:70]: I don't, I didn't change anything in this block of code

File: packages/@ama-sdk/schematics/schematics/index.it.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (kpanot) [reply packages/@ama-sdk/schematics/schematics/index.it.spec.ts]: It is what I meant actually 😁

File: packages/@ama-sdk/schematics/schematics/typescript/core/index.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (kpanot) [reply packages/@ama-sdk/schematics/schematics/typescript/core/index.ts:72]: My bad it was in another function (and I did not noticed the github diff cut)

**Comment by pginoux-1A** [comment]
As discussed, npm scripts prefix swagger could be changed to spec instead. Apart from that everything seems fine to me :)

**Review by kpanot — DISMISSED** [review]

**Review by pginoux-1A — DISMISSED** [review]

**Review by pginoux-1A — APPROVED** [review]

**Review by kpanot — APPROVED** [review]

**Pull request merged**